std/algorithm/sorting.d

```diff
@@ -1048,144 +1048,300 @@ unittest
+struct Merge(alias less = "a < b", Rs...)
+    if (Rs.length >= 2 &&
+        allSatisfy!(isInputRange, Rs) &&
+        !is(CommonType!(staticMap!(ElementType, Rs)) == void))
+{
+    public Rs source;
+    private size_t _lastFrontIndex = size_t.max;
+    static if (isBidirectional)
+    {
+        private size_t _lastBackIndex = size_t.max; // `size_t.max` means uninitialized,
+    }
 
-Params:
-    less = Predicate the given ranges are sorted by.
-    rs = The ranges to compute the union for.
+    import std.functional : binaryFun;
+    import std.traits : isCopyable;
+    import std.typetuple : anySatisfy;
 
-Returns:
-    A range containing the union of the given ranges.
- */
-struct Merge(alias less = "a < b", Rs...) if (allSatisfy!(isInputRange, Rs))
-{
-private:
-    Rs _r;
-    alias comp = binaryFun!(less);
-    uint _crt;
+    private alias comp = binaryFun!less;
+    private alias ElementType = CommonType!(staticMap!(.ElementType, Rs));
+    private enum isBidirectional = allSatisfy!(isBidirectionalRange, staticMap!(Unqual, Rs));
 
-    void adjustPosition(uint candidate = 0)()
+    debug private enum canCheckSortedness = isCopyable!ElementType && !hasAliasing!ElementType;
+
+    this(Rs source)
+    {
+        this.source = source;
+        this._lastFrontIndex = frontIndex;
+    }
+
+    static if (anySatisfy!(isInfinite, Rs))
     {
-        static if (candidate == Rs.length)
+        enum bool empty = false; // propagate infiniteness
+    }
+    else
+    {
+        @property bool empty()
         {
-            _crt = _crt.max;
+            return _lastFrontIndex == size_t.max;
         }
-        else
+    }
+
+    @property auto ref front()
+    {
+        final switch (_lastFrontIndex)
         {
-            if (_r[candidate].empty)
+            foreach (i, _; Rs)
             {
-                adjustPosition!(candidate + 1)();
-                return;
+            case i:
+                assert(!source[i].empty);
+                return source[i].front;
             }
-            foreach (i, U; Rs[candidate + 1 .. $])
-            {
-                enum j = candidate + i + 1;
-                if (_r[j].empty) continue;
-                if (comp(_r[j].front, _r[candidate].front))
-                {
-                    // a new candidate was found
-                    adjustPosition!(j)();
-                    return;
-                }
-            }
-            // Found a successful candidate
-            _crt = candidate;
         }
     }
 
-public:
-    alias ElementType = CommonType!(staticMap!(.ElementType, Rs));
-    static assert(!is(CommonType!(staticMap!(.ElementType, Rs)) == void),
-        typeof(this).stringof ~ ": incompatible element types.");
-
-    ///
-    this(Rs rs)
+    private size_t frontIndex()
     {
-        this._r = rs;
-        adjustPosition();
-    }
-
-    ///
-    @property bool empty()
-    {
-        return _crt == _crt.max;
+        size_t bestIndex = size_t.max; // indicate undefined
```

Comment:
With `ptrdiff_t` you could use `-1` - this seems to be used at a couple of similar bits at Phobos, e.g.
https://github.com/dlang/phobos/blob/master/std/algorithm/searching.d#L881
https://github.com/dlang/phobos/blob/master/std/string.d#L383

Explanation:
http://en.cppreference.com/w/cpp/types/ptrdiff_t